pkg/signer/signer.go

```diff
@@ -0,0 +1,269 @@
+// message, checking for the integrity and authenticity of the provided message in r. It will return
+// the identity of the signer if successful along with the message contents.
+func verifySignatureWithKeyring(r io.Reader, keyring openpgp.EntityList) ([]byte, string, error) {
+	md, err := openpgp.ReadMessage(r, keyring, nil, nil)
+	if err != nil {
+		return nil, "", fmt.Errorf("could not read the message: %s", err)
+	}
+	if !md.IsSigned {
+		return nil, "", fmt.Errorf("not signed")
+	}
+	content, err := ioutil.ReadAll(md.UnverifiedBody)
+	if err != nil {
+		return nil, "", err
+	}
+	if md.SignatureError != nil {
+		return nil, "", fmt.Errorf("signature error: %s", md.SignatureError)
+	}
+	if md.SignedBy == nil {
+		return nil, "", fmt.Errorf("invalid signature")
+	}
+	if md.Signature != nil {
+		if md.Signature.SigLifetimeSecs != nil {
+			expiry := md.Signature.CreationTime.Add(time.Duration(*md.Signature.SigLifetimeSecs) * time.Second)
+			if time.Now().After(expiry) {
+				return nil, "", fmt.Errorf("signature expired on %s", expiry)
+			}
+		}
+	} else if md.SignatureV3 == nil {
+		return nil, "", fmt.Errorf("unexpected openpgp.MessageDetails: neither Signature nor SignatureV3 is set")
+	}
+
+	// follow conventions in containers/image
+	return content, strings.ToUpper(fmt.Sprintf("%x", md.SignedBy.PublicKey.Fingerprint)), nil
+}
+
+// An atomic container signature has the following schema:
+//
+// {
+// 	"critical": {
+// 			"type": "atomic container signature",
+// 			"image": {
+// 					"docker-manifest-digest": "sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e"
+// 			},
+// 			"identity": {
+// 					"docker-reference": "docker.io/library/busybox:latest"
+// 			}
+// 	},
+// 	"optional": {
+// 			"creator": "some software package v1.0.1-35",
+// 			"timestamp": 1483228800,
+// 	}
+// }
+type signature struct {
+	Critical criticalSignature `json:"critical"`
+	Optional optionalSignature `json:"optional"`
+}
+
+type criticalSignature struct {
+	Type     string           `json:"type"`
+	Image    criticalImage    `json:"image"`
+	Identity criticalIdentity `json:"identity"`
+}
+
+type criticalImage struct {
+	DockerManifestDigest string `json:"docker-manifest-digest"`
+}
+
+type criticalIdentity struct {
+	DockerReference string `json:"docker-reference"`
+}
+
+type optionalSignature struct {
+	Creator   string `json:"creator"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+// verifyAtomicContainerSignature verifiers that the provided data authenticates the
+// specified release digest. If error is returned the provided data does NOT authenticate
+// the release digest and the signature must be ignored.
+func verifyAtomicContainerSignature(data []byte, releaseDigest string) error {
+	d := json.NewDecoder(bytes.NewReader(data))
+	d.DisallowUnknownFields()
+	var sig signature
+	if err := d.Decode(&sig); err != nil {
+		return fmt.Errorf("the signature is not valid JSON: %s", err)
+	}
```

Comment:
This is not sufficiently paranoid:
```go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
)

type criticalSignature struct {
	Type string `json:"type"`
}

var malicious = `
{
 "type": "1",
 "type": "2"
}
`

func main() {
	d := json.NewDecoder(bytes.NewReader([]byte(malicious)))
	d.DisallowUnknownFields()
	var sig criticalSignature
	if err := d.Decode(&sig); err != nil {
		fmt.Printf("the signature is not valid JSON: %s\n", err)
	} else {
		fmt.Printf("Successfully parsed as %#v\n", sig)
	}
}
```
parses the input as "type": "2" without rejecting it as ambiguous/invalid.

_Please_ use containers/image/signature, if not all of `PolicyContext`, then at least `VerifyDockerManifestSignature`.  I’m not saying that it is perfect, but it at least allows us to focus the review and testing effort in one place. [This does not have a single unit test…]

(I’m not so insistent about the signing path, because it deals with trusted data and because this project is going to insist on using the Golang-only openpgp implementation.)

---

Or maybe just drop the verification path; AFAICS it has no users in this PR.